x-pack/solutions/security/plugins/security_solution/public/app/links/application_links_updater.ts

```diff
@@ -103,19 +107,26 @@ class ApplicationLinksUpdater {
         return acc;
       }
 
+      // props to be assigned to the current link and its children
+      const extraProps: Partial<LinkItem> = { ...inheritedProps };
+
       if (
-        !hasCapabilities(capabilities, appLinkInfo.capabilities) ||
+        !existCapabilities(capabilities, appLinkInfo.capabilities) ||
         !this.isLinkLicenseAllowed(appLinkInfo, license)
       ) {
-        if (upselling.isPageUpsellable(appLinkInfo.id)) {
-          acc.push({ ...appLinkInfo, unauthorized: true });
+        // The link is not available in the current product payment plan
+        if (!upselling.isPageUpsellable(appLinkInfo.id)) {
+          return acc; // no upselling registered for this link, just exclude it
         }
-        return acc; // not adding sub-links for links that are not authorized
+        extraProps.unavailable = true;
+      } else if (!hasCapabilities(capabilities, appLinkInfo.capabilities)) {
+        // The link is available for the product payment plan but the user doesn't have privileges to access it
+        extraProps.unauthorized = true;
       }
```

Comment:
Technically `filterAppLinks()` isn't just a filter anymore. It enriches app links with `unavailable` and `unauthorized` fields. On top of that it's kind of hard to reason about the behavior here due to multiple `if` conditions. And the shortcut `!upselling.isPageUpsellable(appLinkInfo.id)` makes it hard to extract it to a function though it looks like a good candidate for that.

I'd recommend considering to rename the function to something like `prepareAppLinks` and try to simplify the implementation by decomposing it into smaller functions with readable names to reduce maintainability costs.

Reply:
That's a good point, I changed the function name to `processAppLinks`.

I agree that the `isPageUpsellable` check is somewhat confusing. I tried to move that to the deepLinks creation, but services are not available there, and I think it would be even more confusing, so I left it here.

Added some comments to make it easier to understand the conditions.

https://github.com/elastic/kibana/pull/220552/commits/e625fdb0fe000689c174d1d9950d44a533760f7a